wasmtime/_sharedmemory.py

```diff
@@ -0,0 +1,149 @@
+from . import _ffi as ffi
+from ctypes import *
+import ctypes
+import typing
+from wasmtime import MemoryType, WasmtimeError, Engine
+from ._store import Storelike
+
+
+
+class SharedMemory:
+    _sharedmemory: ffi.wasmtime_sharedmemory_t
+
+    def __init__(self, engine: Engine, ty: MemoryType):
+        """
+        Creates a new shared memory in `store` with the given `ty`
+        """
+
+        sharedmemory = ffi.wasmtime_sharedmemory_t()
+        ptr = POINTER(ffi.wasmtime_sharedmemory_t)()
+        error = ffi.wasmtime_sharedmemory_new(engine.ptr(), ty.ptr(), byref(ptr))
+        if error:
+            raise WasmtimeError._from_ptr(error)
+        self._sharedmemory = sharedmemory
+
+    @classmethod
+    def _from_raw(cls, sharedmemory: ffi.wasmtime_sharedmemory_t) -> "SharedMemory":
+        ty: "SharedMemory" = cls.__new__(cls)
+        ty._sharedmemory = sharedmemory
+        return ty
+
+    def type(self) -> MemoryType:
+        """
+        Gets the type of this memory as a `MemoryType`
+        """
+
+        ptr = ffi.wasmtime_sharedmemory_type(byref(self._sharedmemory))
+        return MemoryType._from_ptr(ptr, None)
+
+    def grow(self, delta: int) -> int:
+        """
+        Grows this memory by the given number of pages
+        """
+
+        if delta < 0:
+            raise WasmtimeError("cannot grow by negative amount")
+        prev = ffi.c_uint64(0)
+        error = ffi.wasmtime_sharedmemory_grow(byref(self._sharedmemory), delta, byref(prev))
+        if error:
+            raise WasmtimeError._from_ptr(error)
+        return prev.value
+
+    def size(self) -> int:
+        """
+        Returns the size, in WebAssembly pages, of this shared memory.
+        """
+
+        return ffi.wasmtime_sharedmemory_size(byref(self._sharedmemory))
+
+    def data_ptr(self) -> "ctypes._Pointer[c_ubyte]":
+        """
+        Returns the raw pointer in memory where this wasm shared memory lives.
+
+        Remember that all accesses to wasm shared memory should be bounds-checked
+        against the `data_len` method.
+        """
+        return ffi.wasmtime_sharedmemory_data(byref(self._sharedmemory))
+
+    def get_buffer_ptr(self, store: Storelike,
+                       size: typing.Optional[int] = None,
+                       offset: int = 0) -> ctypes.Array:
+        """
+        return raw pointer to buffer suitable for creating zero-copy writable NumPy Buffer Protocol
+        this method is also used internally by `read()` and `write()`
+
+        np_mem = np.frombuffer(memory.get_buffer_ptr(store), dtype=np.uint8)
+        np_mem[start:end] = A # write
+        B = np_mem[start:end] # read
+        """
+        if size is None:
+            size = self.data_len(store)
+        ptr_type = ctypes.c_ubyte * size
+        return ptr_type.from_address(ctypes.addressof(self.data_ptr(store).contents) + offset)
+
+    def read(
```

Comment:
This has been copied straight from the Memory code and I didn't put too much attention just yet...

Reply:
I think it looks reasonable enough but you'll be able to remove the `store` argument

Reply:
It may also be worthwhile extracting this to a helper to be shared between shared and non-shared memory?

Reply:
I just realized that the read/write methods and their respective supporting methods, rely on the Store, but the SharedMemory doesn't seem to need a Store. So I wento the Rust API implementation in wasmtime and saw that there are no read/write methods in there, which I guess is aligned with the security model (or a lack of standarization in the IO for shared memory just yet).
I have removed the read/write methdos and their respective supporing methdos.